### PR TITLE
Fix PDB and HPA related issues

### LINKED ISSUE #24000: Validate PDB and HPA settings are compatible
autoscaling enabled causes PDBs to be installed, but there's no enforcement of minReplicas to be compliant with PDB which means that if the value is too low pods can be completely prevented from restarting, blocking rescheduling, upgrade etc. 
This feature would at least validate that the settings are correct.

### PR DESCRIPTION
**Pull request by shamsher31**
Attempt to fix https://github.com/istio/istio/issues/24000 and https://github.com/istio/istio/issues/12602

Some of the goals:

- ~~Increase minReplicas of HPA to 2 https://github.com/istio/istio/issues/24000#issuecomment-633274062 https://github.com/istio/istio/issues/12602#issuecomment-663169633 https://github.com/istio/istio/issues/12602#issuecomment-661487599~~
- Make PDB configurable https://github.com/istio/istio/issues/12602#issuecomment-527770538
- Make sure that HPA and PDB work together.
- Do not apply PDB if HPA is disabled.

More detail:
https://github.com/kubernetes/kubernetes/pull/90253
https://github.com/kubernetes/kubernetes/pull/69867
https://github.com/kubeflow/manifests/issues/1343
https://github.com/gardener/machine-controller-manager/issues/459
https://github.com/istio/istio/issues/23075

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
Not sure what this does? aren't they all set to `minAvailable: 1` in the charts?

**Comment by howardjohn** [comment]
Should we even be expanding the helm api? This is configurable with k8s settings: https://preliminary.istio.io/latest/docs/setup/install/istioctl/#customize-kubernetes-settings

cc @ostromart @costinm

**Comment by shamsher31** [comment]
> I will do the followup PR that will make minAvailable configurable.

WIP: https://github.com/shamsher31/istio/commit/48ec49f30350f7cc8ca611a283113e2ca2ef605d

**Review by richardwxn — COMMENTED** [review]
we can modify from IstioOperator API directly, right? https://github.com/istio/api/blame/master/operator/v1alpha1/operator.proto#L457

**Comment by shamsher31** [comment]
> we can modify from IstioOperator API directly, right? https://github.com/istio/api/blame/master/operator/v1alpha1/operator.proto#L457

Yes, this PR is related to enabling configurable defaults for PDB.

**Comment by shamsher31** [comment]
Related info on setting PDB.

https://github.com/istio/istio/issues/12602#issuecomment-489449657
https://github.com/istio/istio/issues/24000

File: manifests/charts/gateways/istio-egress/templates/poddisruptionbudget.yaml [hunk 1]
```diff
@@ -1,5 +1,5 @@
-{{- if .Values.global.defaultPodDisruptionBudget.enabled }}
 {{ $gateway := index .Values "gateways" "istio-egressgateway" }}
+{{- if and $gateway.autoscaleEnabled .Values.global.defaultPodDisruptionBudget.enabled (ge $gateway.autoscaleMin 2.0) }}
```
Comment (shamsher31) [inline manifests/charts/gateways/istio-egress/templates/poddisruptionbudget.yaml:2]: > (ge $gateway.autoscaleMin 2.0)

This comparison is wired. It should have work with `int` or at least typecast to `int`
https://stackoverflow.com/questions/46009933/kubernetes-helm-chart-if-condition-check?rq=1

**Comment by shamsher31** [comment]
@carolynhu As per our discussion on Slack. I have covered few points mentioned here https://github.com/istio/istio/pull/27209#issue-483774566

**Comment by shamsher31** [comment]
cc: @howardjohn @ostromart @carolynhu

File: manifests/charts/gateways/istio-egress/templates/poddisruptionbudget.yaml (hunk 1 above)
Comment (shamsher31) [reply manifests/charts/gateways/istio-egress/templates/poddisruptionbudget.yaml:2]: Or maybe check `autoscaleMin >= 2` during [translateHPASpec](https://github.com/istio/istio/blob/master/operator/pkg/translate/translate_value.go#L397)

File: manifests/charts/gateways/istio-egress/values.yaml [hunk 2]
```diff
@@ -35,7 +35,7 @@ gateways:
     rollingMaxSurge: 100%
     rollingMaxUnavailable: 25%
     autoscaleEnabled: true
-    autoscaleMin: 1
+    autoscaleMin: 2
```
Comment (howardjohn) [inline manifests/charts/gateways/istio-egress/values.yaml:38]: See my comment on https://github.com/istio/istio/pull/25873#pullrequestreview-455854430. I do not think this is a change we want to be making

Comment (shamsher31) [reply manifests/charts/gateways/istio-egress/values.yaml:38]: Looking at all of the issues and replies related to `PDB` starting from `1.3`, Most of the folks recommend `minReplicas=2` and it solves their problems. https://github.com/istio/istio/issues/12602#issuecomment-663169633 https://github.com/istio/istio/issues/12602#issuecomment-701038118
https://github.com/istio/istio/issues/12602#issuecomment-661487599

How about making `minReplicas` configurable so cluster with resource restriction will set it to lower values?

Comment (howardjohn) [reply manifests/charts/gateways/istio-egress/values.yaml:38]: As mentioned in the comment, I am perfectly happy with allowing it to be higher, making an HA profile, etc. I am not, personally, ok with the default being greater than 1

Comment (howardjohn) [reply manifests/charts/gateways/istio-egress/values.yaml:38]: Put another way - I would rather have no PDB than 2 replicas default. Others may disagree with me though

Comment (shamsher31) [reply manifests/charts/gateways/istio-egress/values.yaml:38]: @sdake WDYT?

Comment (sdake) [reply manifests/charts/gateways/istio-egress/values.yaml:38]: agree with @howardjohn for the default profile. It is simple to add a profile and the profile solves the problem you are after solving. I don't see problems with adding profiles that make sense. The directory is here: https://github.com/istio/istio/tree/master/manifests/profiles

As to how the profile behaves, I don't have a strong opinion. But I think changing the default profile in this way will lead to frustration from smaller operators during the evaluation period.

Cheers,
-steve

File: operator/cmd/mesh/testdata/manifest-generate/data-snapshot/charts/istio-control/istio-discovery/files/gen-istio.yaml [hunk 3]
```diff
@@ -867,7 +867,7 @@ metadata:
     istio.io/rev: default
 spec:
   maxReplicas: 5
-  minReplicas: 1
+  minReplicas: 2
```
Comment (carolynhu) [inline operator/cmd/mesh/testdata/manifest-generate/data-snapshot/charts/istio-control/istio-discovery/files/gen-istio.yaml:870]: we can keep this minReplicas:1

Comment (shamsher31) [reply operator/cmd/mesh/testdata/manifest-generate/data-snapshot/charts/istio-control/istio-discovery/files/gen-istio.yaml]: Reverted.

File: manifests/charts/gateways/istio-egress/templates/poddisruptionbudget.yaml [hunk 4]
```diff
@@ -1,5 +1,5 @@
-{{- if .Values.global.defaultPodDisruptionBudget.enabled }}
 {{ $gateway := index .Values "gateways" "istio-egressgateway" }}
+{{- if and $gateway.autoscaleEnabled .Values.global.defaultPodDisruptionBudget.enabled }}
```
Comment (bianpengyuan) [inline manifests/charts/gateways/istio-egress/templates/poddisruptionbudget.yaml:2]: Should we check if pod disruption budget min available is less than hpa min replica before setting pdb?

**Comment by bianpengyuan** [comment]
Ping on review for this @sdake @howardjohn This addresses a P0 for 1.8 so would be great to check it in asap.

File: manifests/charts/gateways/istio-egress/templates/poddisruptionbudget.yaml (hunk 4 above)
Comment (bianpengyuan) [reply manifests/charts/gateways/istio-egress/templates/poddisruptionbudget.yaml:2]: Or should we add some check into istioctl to make sure user does not make invalid config with pdb+hpa?

**Comment by shamsher31** [comment]
Closing in favour of https://github.com/istio/istio/pull/28180

**Pull request closed without merging**